File: 167. Two Sum II - Input array is sorted/binary-search.cpp
// Time : O(nlogn)
// Space : O(1)

class Solution {
private:
    // predicate : a <= b
    bool p(int a, int b) {
        return a <= b;
    }
    
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        for (int i = 0; i < numbers.size(); i++) {
            int lo = i + 1;
            int hi = numbers.size() - 1;
            while (lo < hi) {
                int mid = lo + (hi - lo) / 2;
                if (p(target-numbers[i], numbers[mid])) {
                    hi = mid;
                }
                else {
                    lo = mid + 1;
                }
            }
            
            if (numbers[i] + numbers[lo] == target) {
                return vector<int> {i + 1, lo + 1};
            }
        }
    }
};

File: 167. Two Sum II - Input array is sorted/binary-search.py
class Solution:
    def twoSum(self, numbers, target):
        """
        :type numbers: List[int]
        :type target: int
        :rtype: List[int]
        """
        solution = []
        for i in range(len(numbers)):
            real_target = target - numbers[i]
            index = binary_search(numbers, i + 1, real_target)
            if index != -1:
                solution.append(i + 1)
                solution.append(index + 1)
                return solution
    

def binary_search(array, lo, target):
    hi = len(array) - 1
    while lo <= hi:
        mid = (lo + hi) // 2
        if array[mid] == target:
            return mid
        elif array[mid] > target:
            hi = mid - 1
        else:
            lo = mid + 1
    
    return -1